File: sdks/typescript/aion-client/examples/seven-operations.ts
import {
  AlreadyExistsError,
  QueryTimeoutError,
  UnavailableError,
  connect,
} from "../src/index.js";

type EchoState = {
  readonly lastSignal?: string;
};

async function run(): Promise<void> {
  const endpoint =
    process.argv[2] ?? process.env.AION_SERVER_URL ?? "http://127.0.0.1:8080";
  const bearerToken = process.env.AION_AUTH_TOKEN;
  const client = await connect({
    endpoint,
    namespace: "conformance",
    auth:
      bearerToken === undefined
        ? undefined
        : { bearerToken, namespaces: ["conformance"] },
    tls: {
      enabled: process.env.AION_INSECURE !== "1" && endpoint.startsWith("https://"),
    },
  });

  const handle = await client.start({
    workflowType: "conformance.echo",
    input: { message: "hello", counter: 1 },
    idempotencyKey: `aion-client-typescript-seven-operations-${process.pid}`,
  });
  console.log(
    `started workflow=${handle.workflowId} run=${handle.runId ?? "latest"}`,
  );

  await handle.signal({
    signalName: "record",
    payload: { value: "signal-observed" },
  });
  console.log("sent signal record");

  const state = await handle.query<undefined, EchoState>({
    queryName: "state",
    timeoutMs: 5_000,
  });
  console.log("query state", state);

  const { summaries } = await client.list({ namespace: "conformance" });
  console.log(`listed ${summaries.length} workflow(s)`);

  const description = await handle.describe({ includeHistory: true });
  console.log(`described ${description.history.length} event(s)`);

  await handle.cancel({
    reason: "seven-operations example requested cancellation",
  });
  console.log("cancel requested");

  for await (const event of handle.subscribe({ maxReconnects: 3 })) {
    console.log(`subscribed event seq=${event.seq}`);
    break;
  }
}

try {
  await run();
} catch (error) {
  if (error instanceof UnavailableError) {
    console.error(
      "aion-server is unavailable; check AION_SERVER_URL and the fixture",
    );
  } else if (error instanceof AlreadyExistsError) {
    console.error("idempotency key was reused for a different start request");
  } else if (error instanceof QueryTimeoutError) {
    console.error("query timed out before the fixture replied");
  } else {
    throw error;
  }
  process.exitCode = 1;
}

File: sdks/python/aion-client/examples/seven_operations.py
"""Runnable aion-client-python example covering all seven workflow operations."""

from __future__ import annotations

import asyncio
import os
import sys
from typing import Any, NoReturn, cast

from aion_client import AlreadyExists, Client, EventStream, QueryTimeout, StreamEvent, TLSConfig, Unavailable


def _endpoint() -> str:
    if len(sys.argv) > 1:
        return sys.argv[1]
    return os.environ.get("AION_SERVER_URL", "http://127.0.0.1:50051")


def _fail(message: str) -> NoReturn:
    print(message, file=sys.stderr)
    raise SystemExit(1)


def _tls_enabled(endpoint: str) -> bool:
    if os.environ.get("AION_INSECURE") == "1":
        return False
    return endpoint.startswith("https://") or endpoint.startswith("grpcs://")


async def run() -> None:
    endpoint = _endpoint()
    async with await Client.connect(
        endpoint,
        auth=os.environ.get("AION_AUTH_TOKEN"),
        tls=TLSConfig(enabled=_tls_enabled(endpoint)),
        namespace="conformance",
        # The WebSocket event stream rides the server's HTTP listener — a
        # separate address from the gRPC endpoint; there is no default.
        stream_endpoint=os.environ.get("AION_STREAM_URL"),
    ) as client:
        handle = await client.start(
            "conformance_echo",
            {"message": "hello", "counter": 1},
            idempotency_key=f"aion-client-python-seven-operations-{os.getpid()}",
        )
        print(f"started workflow={handle.workflow_id} run={handle.run_id}")

        await handle.signal("record", {"value": "signal-observed"})
        print("sent signal record")

        state = cast(dict[str, Any], await handle.query("state", target_type=dict, timeout=5.0))
        print(f"query state={state}")

        summaries = await client.list()
        print(f"listed {len(summaries)} workflow(s)")

        description = await handle.describe(include_history=True)
        print(f"described {len(description.history)} event(s)")

        await handle.cancel(reason="seven-operations example requested cancellation")
        print("cancel requested")

        stream: EventStream[Any] = handle.subscribe(raw=True)
        try:
            event = cast(StreamEvent[Any], await asyncio.wait_for(stream.__anext__(), timeout=5.0))
            print(f"subscribed event seq={event.seq}")
        except asyncio.TimeoutError:
            raise QueryTimeout("timed out waiting for the first subscribed event") from None
        finally:
            await stream.aclose()


async def main() -> None:
    try:
        await run()
    except Unavailable:
        _fail("aion-server is unavailable; check AION_SERVER_URL and the fixture")
    except AlreadyExists:
        _fail("idempotency key was reused for a different start request")
    except QueryTimeout:
        _fail("query or subscribe timed out before the fixture replied")


if __name__ == "__main__":
    asyncio.run(main())
